### PR TITLE
#KM-20/ Set up TinaCMS

### PR DESCRIPTION
**Pull request by avinashi10**
## Description
This PR sets up the configuration of TinaCMS, including schema to model the data that the CMS will hold, a visual editor to edit content, and queries to import the content to the app. The header and footer are not yet connected to the CMS.

## Related Tickets & Documents
Closes #KM-20

## Acceptance Criteria

- [x] a CMS has been researched and selected for ease of use for developers and robust visual editing for clients
- [x] the CMS data modeling matches the website content
- [x] the app queries the CMS to get real-time content

## What type of PR is this? (check all applicable)

|     | Type                       |
| --- | -------------------------- |
|  ✓ | :sparkles: New feature     |
|     | 🎨 Style                   |
|  ✓ | :hammer: Refactoring       |
|     | 🔥 Performance Improvements|
|     | :bug: Bug fix              |
|     | :100: Add tests            |
|  ✓ | :link: Update dependencies |
|     | :scroll: Docs              |

## Mobile & Desktop Screenshots/Recordings
Screenshot 2024-02-02 at 3 52 22 PM

## What I learned
TinaCMS has real-time editing capabilities and what's often called a 'WYSIWYG' (what you see is what you get) interface. In development we'll run TinaCMS locally, but in production we'll use the self-hosted option of Tina Cloud.

## Testing steps

1. `git switch KM-20/feature-CMS`
2. `npm i` - so TinaCMS is installed!
3. `npm run dev` after installing
4. navigate to localhost:3000/admin (be sure to include the slug or the page will only load the header and footer) NOTE: it's a minute for the page to load, and another beat for the components to load
5. Once everything is loaded, you should see the visual editor on the left. It should be titled "Pages" and subtitled "blocks". Each block represents a separate component on the home page. Click into each block and play around with the text, you should see it change in real time on the browser screen on the right. Simply hit 'Reset' to undo any changes. Changes are really easy to make, so simply change something back if you accidentally saved, or want to see how long saving takes (it's fast!)
6. Feel free to check out the code to look into the setup, it'll be easiest to explain if I walk you through it!

## Added to documentation?

|     | Type                       |
| --- | -------------------------- |
|     | 📜 README.md    |
| ✓  | 🙅 no documentation needed    |

## What gif best describes this PR?
  woohoo

### CODE REVIEW

File: tina/config.ts
Comment (avinashi10) [inline tina/config.ts:1]: This file is auto-generated by initializing tina in your repo - with the option to initialize for nextjs/ts

File: tina/config.ts [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+  media: {
+    tina: {
+      mediaRoot: "",
+      publicFolder: "public",
+    },
+  },
+  // See docs on content modeling for more info on how to setup new content models: https://tina.io/docs/schema/
+  // tina/config.ts
+  schema: {
+    collections: [
```
Comment (avinashi10) [inline tina/config.ts:31]: This is the customizable part of this file which is where you create the schema for how you want the data fields to show up in the editor

File: content/pages/home.md
Comment (avinashi10) [inline content/pages/home.md:1]: When you add content via the CMS, it generates a file (either markdown or json, defaults to markdown) with all that data.

File: app/page.tsx
Comment (avinashi10) [inline app/page.tsx:1]: Once the schema is set up and the content is added via the CMS (and the file containing the content autogenerated), you query the file using built-in graphQL queries.

File: app/page.tsx [hunk 2]
```diff
@@ -1,27 +1,12 @@
...
-import Image from "next/image";
+// LOCAL IMPORTS
+import Home from "./components/Home";
 
-//LOCAL IMPORTS
-import Hero from "./components/Hero";
-import CompanyValues from "./components/CompanyValues";
+export default async function Page() {
+  // CMS QUERIES
+  const result = await client.queries.page({ relativePath: "home.md" });
```
Comment (avinashi10) [inline app/page.tsx:9]: This syntax is copied and pasted from TinaCMS docs, client is a built-in function that we import, queries is one of its properties, and page is the name we gave our collection (refer to schema in config)

File: app/page.tsx [hunk 3]
```diff
@@ -1,27 +1,12 @@
...
-    fontSize: "16px",
-  };
-
-  return (
-    <Grid templateColumns={"1fr"}>
-      <Hero />
-      <CompanyValues />
-    </Grid>
-  );
+  return <Home {...result} />;
```
Comment (avinashi10) [inline app/page.tsx:11]: I had to refactor the code and create an extra outer 'layer' bc the data query is async, and that can't happen in client-side components. So I made this a server-side component, queried the data, and passed all the data as props. The Home component has now moved to the components folder.

File: pages/demo/blog/[filename].tsx [hunk 4]
```diff
@@ -0,0 +1,218 @@
+// THIS FILE HAS BEEN GENERATED WITH THE TINA CLI.
```
Comment (BrettEastman) [inline pages/demo/blog/[filename].tsx]: This is a demo file and should be deleted before merging to main.

File: public/kenworthy-logo.jpg
Comment (BrettEastman) [inline public/kenworthy-logo.jpg:1]: This is a duplicate. The same file already exists in public/images.

File: public/machine-gear-background.png
Comment (BrettEastman) [inline public/machine-gear-background.png:1]: This is a duplicate. The same file already exists in public/images.

File: public/GaneshCyclone.png
Comment (BrettEastman) [inline public/GaneshCyclone.png:1]: Perhaps all images should be stored in public/images unless there is a specific reason not to.

File: tina/config.ts [hunk 5]
```diff
@@ -0,0 +1,196 @@
+import { defineConfig } from "tinacms";
+
+// Your hosting provider likely exposes this as an environment variable
+const branch =
+  process.env.GITHUB_BRANCH ||
```
Comment (BrettEastman) [inline tina/config.ts:5]: If there are .env codes we need to get from you, please send them securely. Thanks!

File: app/components/CompanyValues.tsx [hunk 6]
```diff
@@ -46,22 +48,22 @@ export default function CompanyValues() {
...
-                </Text>
-                <Text fontSize="lg">
-                  {
-                    "Our team is dedicated to making sure your parts are correct an in tolerance every time. They use their expertise, as well as modern inspection equipment to achieve this goal. Upon request we can provide inspection reports for your parts with detailed measurements compared to drawing specifications. We also offer laser engraving to give your parts a professional appearance."
-                  }
-                </Text>
+                <TinaMarkdown
+                  content={props.quality1?.description}
+                  components={{
+                    p: ({ children }) => (
```
Comment (BrettEastman) [inline app/components/CompanyValues.tsx]: Just a few Typescript errors we can fix together.

File: app/components/CompanyValues.tsx [hunk 7]
```diff
@@ -46,22 +48,22 @@ export default function CompanyValues() {
...
+                      <Text as="span" fontSize="xl">
```
Comment (BrettEastman) [inline app/components/CompanyValues.tsx]: The text is showing up as white against a white background in my browser.

File: app/components/Home.tsx [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+  data: PageQuery;
+  variables: {
+    relativePath: string;
+  };
+  query: string;
+}) {
+  // HOOKS
+  const { data } = useTina(props);
+
+  // console.log("data props:", data);
```
Comment (BrettEastman) [inline app/components/Home.tsx]: Remove unneeded console logs before merging to main.

File: app/components/Home.tsx [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    <Grid templateColumns={"1fr"}>
+      {data.page.blocks?.map((block, i) => {
+        switch (block?.__typename) {
+          case "PageBlocksHero": {
+            return <Hero key={i} {...block} />;
+          }
+          case "PageBlocksValues": {
+            return <CompanyValues key={i} {...block} />;
+          }
+          // case "PageBlocksLocation": {
```
Comment (BrettEastman) [inline app/components/Home.tsx:32]: Are these future cases for this switch case statement? If so, delete now if not needed for this PR then bring them back later on an update. If there are only 2-3 cases, you could also consider refactoring it to an if block for clarity, but not necessary.

File: app/page.tsx (hunk 3 above)
Comment (BrettEastman) [reply app/page.tsx:11]: Sounds good!

File: public/kenworthy-logo.jpg
Comment (avinashi10) [reply public/kenworthy-logo.jpg:1]: So when you upload an image to TinaCMS, it automatically saves it to public (vs public/images) and then saves the image url in the markdown file with that path. I think saving in public/images would result in needing to manually change the `src` url in code, which wouldn't work for clients. However, I do need to figure out how images work with TinaCMS in production in self-hosting vs TinaCloud (I think self-hosting with images is less straight-forward)

File: public/machine-gear-background.png
Comment (avinashi10) [reply public/machine-gear-background.png:1]: See comment above

File: public/GaneshCyclone.png
Comment (avinashi10) [reply public/GaneshCyclone.png:1]: See comment above

File: tina/config.ts (hunk 5 above)
Comment (avinashi10) [reply tina/config.ts:5]: There are no variables yet as this login/authentication only runs in production and we're running locally. We can look into production set up when we're closer to that step.

File: app/components/Home.tsx (hunk 9 above)
Comment (avinashi10) [reply app/components/Home.tsx:32]: There's a block for each section of the home page, so there are still 4 more: Location, Our Machines, Request a Quote, Join Our Team (6 total)

File: app/components/CompanyValues.tsx (hunk 7 above)
Comment (avinashi10) [reply app/components/CompanyValues.tsx]: I added a color prop and it updated from Chakra's default black to our custom text color on my browser, can you pull changes and check on yours?

File: public/kenworthy-logo.jpg
Comment (BrettEastman) [reply public/kenworthy-logo.jpg:1]: Alternatively, we could have all the images saved in public and do away with the public/images folder. Whatever TinaCMS research leads you to a solution where all images are stored in one folder and there are no duplicates, I am for it.

File: app/components/CompanyValues.tsx (hunk 7 above)
Comment (BrettEastman) [reply app/components/CompanyValues.tsx]: That fixed it. It looks great now!

**Review by BrettEastman — APPROVED** [review]
@avinashi10, great find with TinaCMS! I had no trouble at all getting it up and running and editing files in the admin mode. Really excellent ease of use for the client and intuitive UI. Once the last couple conversations resolved, approved for merge.

**Pull request merged**